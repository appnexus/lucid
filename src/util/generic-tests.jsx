import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
