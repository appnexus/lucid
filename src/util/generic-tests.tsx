import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
