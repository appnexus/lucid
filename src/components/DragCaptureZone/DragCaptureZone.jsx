import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass } from '../../util/component-definition';

const boundClassNames = lucidClassNames.bind('&-DragCaptureZone');
const {
	func,
	string
} = React.PropTypes;

/**
 * {"categories": ["utility"]}
 *
 * This is a helper component used to capture mouse events to determine when the
 * user starts, is and stops dragging.
 */
const DragCaptureZone = createClass({
	displayName: 'DragCaptureZone',
	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Called as the user drags the mouse.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDrag: func,

		/**
		 * Called when the user releases the mouse button after having dragged.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragEnd: func,

		/**
		 * Called when the user presses the mouse button down while over the
		 * component.
		 *
		 * Signature: `({ dx, dy, pageX, pageY }, { event, props }) => {}`
		 */
		onDragStart: func
	},

	getDefaultProps() {
		return {
			onDrag: _.noop,
			onDragEnd: _.noop,
			onDragStart: _.noop
		};
	},

	getInitialState() {
		return {
			pageX: 0,
			pageY: 0
		};
	},

	render() {
		return (
			<div
				{...this.props}
				className={boundClassNames('&', this.props.className)}
				key='DragCaptureZone'
				onMouseDown={this.handleDragStart}
			/>
		);
	},

	componentWillUnmount() {
		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);
	},

	handleDrag(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		this.props.onDrag({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});
	},

	handleDragEnd(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);

		this.props.onDragEnd({
			dX: pageX - this.state.pageX,
			dY: pageY - this.state.pageY,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX: 0,
			pageY: 0
		});
	},

	handleDragStart(event) {
		const {
			pageX,
			pageY
		} = event;

		event.preventDefault();

		window.document.addEventListener('mousemove', this.handleDrag);
		window.document.addEventListener('mouseup', this.handleDragEnd);

		this.props.onDragStart({
			dX: 0,
			dY: 0,
			pageX,
			pageY
		}, {
			event,
			props: this.props
		});

		this.setState({
			pageX,
			pageY
		});
	}
});

export default DragCaptureZone;
