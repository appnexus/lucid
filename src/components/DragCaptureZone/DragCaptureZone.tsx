import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import {
	omitProps,
	StandardProps,
	Overwrite,
} from '../../util/component-types';

const cx = lucidClassNames.bind('&-DragCaptureZone');
const { func, string } = PropTypes;

export interface IDragCaptureZonePropsRaw extends StandardProps {
	/** Called as the user drags the mouse. */
	onDrag: (
		{
			dX,
			dY,
			pageX,
			pageY,
		}: {
			dX: number;
			dY: number;
			pageX: number;
			pageY: number;
		},
		{
			event,
			props,
		}: {
			event: MouseEvent | TouchEvent;
			props: IDragCaptureZoneProps;
		}
	) => void;

	/** Called when the user releases the mouse button after having dragged. */
	onDragEnd: (
		{
			dX,
			dY,
			pageX,
			pageY,
		}: {
			dX: number;
			dY: number;
			pageX: number;
			pageY: number;
		},
		{
			event,
			props,
		}: {
			event: MouseEvent | TouchEvent;
			props: IDragCaptureZoneProps;
		}
	) => void;

	/** Called when the user presses the mouse button down while over the component. */
	onDragStart: (
		{
			dX,
			dY,
			pageX,
			pageY,
		}: {
			dX: number;
			dY: number;
			pageX: number;
			pageY: number;
		},
		{
			event,
			props,
		}: {
			event:
				| React.MouseEvent<HTMLDivElement, MouseEvent>
				| React.TouchEvent<HTMLDivElement>;
			props: IDragCaptureZoneProps;
		}
	) => void;

	/** Called when the drag event is canceled due to user interaction.
	 * For example: if a system alert pops up during a touch event. */
	onDragCancel: ({
		event,
		props,
	}: {
		event: MouseEvent | TouchEvent;
		props: IDragCaptureZoneProps;
	}) => void;
}

export type IDragCaptureZoneProps = Overwrite<
	React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
	IDragCaptureZonePropsRaw
>;

interface IDragCaptureZoneState {
	pageX: number;
	pageY: number;
}

class DragCaptureZone extends React.Component<
	IDragCaptureZoneProps,
	IDragCaptureZoneState,
	{}
> {
	static displayName = 'DragCaptureZone';
	static peek = {
		description: `This is a helper component used to capture mouse events to determine when the user starts, is and stops dragging.`,
		categories: ['utility'],
	};
	static propTypes = {
		/**
			Appended to the component-specific class names set on the root element.
		*/
		className: string,

		/**
			Called as the user drags the mouse.  Signature:
			\`({ dx, dy, pageX, pageY }, { event, props }) => {}\`
		*/
		onDrag: func,

		/**
			Called when the user releases the mouse button after having dragged.
			Signature: \`({ dx, dy, pageX, pageY }, { event, props }) => {}\`
		*/
		onDragEnd: func,

		/**
			Called when the user presses the mouse button down while over the
			component.  Signature:
			\`({ dx, dy, pageX, pageY }, { event, props }) => {}\`
		*/
		onDragStart: func,

		/**
			Called when the drag event is canceled due to user interaction.  For
			example: if a system alert pops up during a touch event.  Signature:
			\`({ event, props }) => {}\`
		*/
		onDragCancel: func,
	};

	private elementRef = React.createRef<HTMLDivElement>();

	state = {
		pageX: 0,
		pageY: 0,
	};

	static defaultProps = {
		onDrag: _.noop,
		onDragEnd: _.noop,
		onDragStart: _.noop,
		onDragCancel: _.noop,
	};

	handleDrag = (event: MouseEvent | TouchEvent): void => {
		let pageX;
		let pageY;

		/* istanbul ignore next */
		if ('touches' in event) {
			pageX = event.touches[0].pageX;
			pageY = event.touches[0].pageY;
		} else {
			pageX = event.pageX;
			pageY = event.pageY;
		}

		event.preventDefault();

		this.props.onDrag(
			{
				dX: pageX - this.state.pageX,
				dY: pageY - this.state.pageY,
				pageX,
				pageY,
			},
			{
				event,
				props: this.props,
			}
		);
	};

	handleMouseDragStart = (
		event: React.MouseEvent<HTMLDivElement, MouseEvent>
	): void => {
		const pageX = event.pageX;
		const pageY = event.pageY;

		window.document.addEventListener('mousemove', this.handleDrag);
		window.document.addEventListener('mouseup', this.handleDragEnd);

		event.preventDefault();

		this.props.onDragStart(
			{
				dX: 0,
				dY: 0,
				pageX,
				pageY,
			},
			{
				event,
				props: this.props,
			}
		);

		this.setState({
			pageX,
			pageY,
		});
	};

	handleTouchDragStart = (event: React.TouchEvent<HTMLDivElement>): void => {
		const pageX = event.touches[0].pageX;
		const pageY = event.touches[0].pageY;

		event.preventDefault();

		this.props.onDragStart(
			{
				dX: 0,
				dY: 0,
				pageX,
				pageY,
			},
			{
				event,
				props: this.props,
			}
		);

		this.setState({
			pageX,
			pageY,
		});
	};

	handleDragEnd = (event: MouseEvent | TouchEvent): void => {
		let pageX;
		let pageY;

		/* istanbul ignore next */
		if ('changedTouches' in event) {
			pageX = event.changedTouches[0].pageX;
			pageY = event.changedTouches[0].pageY;
		} else {
			pageX = event.pageX;
			pageY = event.pageY;
		}

		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);

		event.preventDefault();

		this.props.onDragEnd(
			{
				dX: pageX - this.state.pageX,
				dY: pageY - this.state.pageY,
				pageX,
				pageY,
			},
			{
				event,
				props: this.props,
			}
		);

		this.setState({
			pageX: 0,
			pageY: 0,
		});
	};

	handleDragCancel = (event: MouseEvent | TouchEvent): void => {
		this.props.onDragCancel({
			event,
			props: this.props,
		});

		this.setState({
			pageX: 0,
			pageY: 0,
		});
	};

	componentDidMount(): void {
		//add event listeners directly on the DOM element to allow preventDefault
		//calls which are not honored due to react's event delegation
		//reference: https://github.com/facebook/react/issues/8968

		if (this.elementRef.current) {
			this.elementRef.current.addEventListener('touchmove', this.handleDrag);
			this.elementRef.current.addEventListener('touchend', this.handleDragEnd);
			this.elementRef.current.addEventListener(
				'touchcancel',
				this.handleDragCancel
			);
		}
	}

	componentWillUnmount(): void {
		if (this.elementRef.current) {
			this.elementRef.current.removeEventListener('touchmove', this.handleDrag);
			this.elementRef.current.removeEventListener(
				'touchend',
				this.handleDragEnd
			);
			this.elementRef.current.removeEventListener(
				'touchcancel',
				this.handleDragCancel
			);
		}
		window.document.removeEventListener('mousemove', this.handleDrag);
		window.document.removeEventListener('mouseup', this.handleDragEnd);
	}

	render(): React.ReactNode {
		return (
			<div
				{...omitProps(this.props, undefined, _.keys(DragCaptureZone.propTypes))}
				className={cx('&', this.props.className)}
				key='DragCaptureZone'
				onMouseDown={this.handleMouseDragStart}
				onTouchStart={this.handleTouchDragStart}
				ref={this.elementRef}
			/>
		);
	}
}

export default DragCaptureZone;
