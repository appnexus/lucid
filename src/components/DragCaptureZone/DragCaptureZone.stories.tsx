import _ from 'lodash';
import React from 'react';
import createClass from 'create-react-class';
import DragCaptureZone from './DragCaptureZone';

export default {
	title: 'Utility/DragCaptureZone',
	component: DragCaptureZone,
	parameters: {
		docs: {
			description: {
				component: (DragCaptureZone as any).peek.description,
			},
		},
	},
};

/* Interactive */
export const Interactive = () => {
	const Component = createClass({
		getInitialState() {
			return {
				events: [],
			};
		},

		render() {
			const { events } = this.state;

			return (
				<section
					style={{
						alignItems: 'center',
						display: 'flex',
					}}
				>
					<DragCaptureZone
						onDrag={this.handleDragged}
						onDragEnd={this.handleDragEnded}
						onDragStart={this.handleDragStarted}
					>
						<div
							style={{
								alignItems: 'center',
								backgroundColor: '#b7b7b7',
								display: 'flex',
								fontFamily: 'Helvetica, sans-serif',
								fontSize: '24px',
								fontWeight: 300,
								height: 300,
								justifyContent: 'center',
								textTransform: 'uppercase',
								width: 400,
							}}
						>
							Go wild!
						</div>
					</DragCaptureZone>
					<div
						style={{
							height: 300,
							marginLeft: 50,
							overflow: 'auto',
							width: 600,
						}}
					>
						{_.reverse(
							_.map(events, ({ coordinates, type }, index) => (
								<div key={index}>
									<div
										style={{
											fontWeight: 'bold',
										}}
									>
										{type}
									</div>
									<div>
										{`dx: ${coordinates.dX}, dy: ${coordinates.dY},
									px: ${coordinates.pageX}, py: ${coordinates.pageY}`}
									</div>
								</div>
							))
						)}
					</div>
				</section>
			);
		},

		handleDragEnded(coordinates: any) {
			this.setState({
				events: _.concat(this.state.events, {
					type: 'end',
					coordinates,
				}),
			});
		},

		handleDragStarted(coordinates: any) {
			this.setState({
				events: _.concat(this.state.events, {
					type: 'start',
					coordinates,
				}),
			});
		},

		handleDragged(coordinates: any) {
			const lastEvent: any = _.last(this.state.events);
			const alreadyDragging = lastEvent.type === 'start';

			this.setState({
				events: _.concat(
					alreadyDragging
						? this.state.events
						: _.slice(this.state.events, 0, -1),
					{
						type: 'drag',
						coordinates,
					}
				),
			});
		},
	});

	return <Component />;
};
Interactive.storyName = 'Interactive';
