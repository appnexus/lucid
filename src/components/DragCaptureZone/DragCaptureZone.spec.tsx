import _ from 'lodash';
import assert from 'assert';
import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';

import { common } from '../../util/generic-tests';

import DragCaptureZone from './DragCaptureZone';

describe('DragCaptureZone', () => {
	common(DragCaptureZone);

	describe('props', () => {
		describe('onDrag', () => {
			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<DragCaptureZone />);

				assert.equal(wrapper.prop('onDrag'), _.noop);
			});
		});

		describe('onDragEnd', () => {
			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<DragCaptureZone />);

				assert.equal(wrapper.prop('onDragEnd'), _.noop);
			});
		});

		describe('onDragStart', () => {
			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<DragCaptureZone />);

				assert.equal(wrapper.prop('onDragStart'), _.noop);
			});
		});

		describe('onDragCancel', () => {
			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<DragCaptureZone />);

				assert.equal(wrapper.prop('onDragCancel'), _.noop);
			});
		});

		describe('pass throughs', () => {
			it('passes through all props to the root element.', () => {
				const wrapper = mount(<DragCaptureZone foo={1} bar={2} baz={3} />);
				const rootProps = wrapper.find('div').props();

				_.forEach(['foo', 'bar', 'baz'], prop => {
					assert(_.has(rootProps, prop));
				});
			});
		});
	});
});

describe('DragCaptureZone', () => {
	const dragCoordinates = {
		pageX: 75,
		pageY: 25,
	};
	const dragEndCoordinates = {
		pageX: 25,
		pageY: 50,
	};
	const dragStartCoordinates = {
		pageX: 50,
		pageY: 75,
	};

	describe('user presses the mouse button down over the rendered component', () => {
		let onDragStart;

		beforeEach(() => {
			onDragStart = sinon.spy();

			mount(<DragCaptureZone onDragStart={onDragStart} />)
				.find('div')
				.simulate('mousedown', dragStartCoordinates);
		});

		it('calls the function passed in as the `onDragStart` prop...', () => {
			assert(onDragStart.calledOnce);
		});

		it('...passes along a map of coordinate data as the first argument...', () => {
			const expectedData = _.assign(
				{
					dX: 0,
					dY: 0,
				},
				dragStartCoordinates
			);

			assert.deepEqual(_.first(onDragStart.args[0]), expectedData);
		});

		it('...and passes along a React synthetic event as part of the second argument.', () => {
			assert(_.last(onDragStart.args[0]).event);
		});
	});

	describe('user moves the mouse button after having started a drag interaction', () => {
		let event;
		let onDrag;

		beforeEach(() => {
			event = document.createEvent('Event');
			_.assign(event, dragCoordinates);
			event.initEvent('mousemove', true, true);

			onDrag = sinon.spy();

			mount(<DragCaptureZone onDrag={onDrag} />)
				.find('div')
				.simulate('mousedown', dragStartCoordinates);
			window.document.dispatchEvent(event);
		});

		it('calls the function passed in as the `onDrag` prop...', () => {
			assert(onDrag.calledOnce);
		});

		it('...passes along a map of coordinate data as the first argument...', () => {
			const expectedData = _.assign(
				{
					dX: dragCoordinates.pageX - dragStartCoordinates.pageX,
					dY: dragCoordinates.pageY - dragStartCoordinates.pageY,
				},
				dragCoordinates
			);

			assert.deepEqual(_.first(onDrag.args[0]), expectedData);
		});

		it('...and passes along a native event as part of the second argument.', () => {
			assert(_.last(onDrag.args[0]).event instanceof window.Event);
		});
	});

	describe('user releases the mouse button after having started a drag interaction', () => {
		let event;
		let onDragEnd;

		beforeEach(() => {
			event = document.createEvent('Event');
			_.assign(event, dragEndCoordinates);
			event.initEvent('mouseup', true, true);

			onDragEnd = sinon.spy();

			mount(<DragCaptureZone onDragEnd={onDragEnd} />)
				.find('div')
				.simulate('mousedown', dragStartCoordinates);
			window.document.dispatchEvent(event);
		});

		it('calls the function passed in as the `onDragEnd` prop...', () => {
			assert(onDragEnd.calledOnce);
		});

		it('...passes along a map of coordinate data as the first argument...', () => {
			const expectedData = _.assign(
				{
					dX: dragEndCoordinates.pageX - dragStartCoordinates.pageX,
					dY: dragEndCoordinates.pageY - dragStartCoordinates.pageY,
				},
				dragEndCoordinates
			);

			assert.deepEqual(_.first(onDragEnd.args[0]), expectedData);
		});

		it('...and passes along a native event as part of the second argument.', () => {
			assert(_.last(onDragEnd.args[0]).event instanceof window.Event);
		});
	});

	describe('the drag action is canceled', () => {
		const mockEvent = {};
		let onDragCancel;

		beforeEach(() => {
			onDragCancel = sinon.spy();

			const wrapper = shallow(<DragCaptureZone onDragCancel={onDragCancel} />, {
				disableLifecycleMethods: true,
			});
			wrapper.instance().handleDragCancel(mockEvent);
		});

		it('calls the function passed in as the `onDragCancel` prop...', () => {
			assert(onDragCancel.calledOnce);
		});

		it('...and passes along a React synthetic event as part of the last argument.', () => {
			assert(_.last(onDragCancel.args[0]).event);
		});
	});
});
