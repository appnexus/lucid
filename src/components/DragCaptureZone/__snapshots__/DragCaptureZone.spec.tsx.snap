// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragCaptureZone [common] example testing should match snapshot(s) for interactive 1`] = `
<div
  className="lucid-DragCaptureZone"
  key="DragCaptureZone"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#b7b7b7",
        "display": "flex",
        "fontFamily": "Helvetica, sans-serif",
        "fontSize": "24px",
        "fontWeight": 300,
        "height": 300,
        "justifyContent": "center",
        "textTransform": "uppercase",
        "width": 400,
      }
    }
  >
    Go wild!
  </div>
</div>
`;
